File: ED del diego/Java/Febrero 2019/src/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2; // Me sobraría

public class Kruskal {

	private static <V> V getRep(V vertex, Dictionary<V,V> dict){
		V v = vertex;
		while(v != dict.valueOf(v)){
			v = dict.valueOf(v);
		}
		return v;
	}

	private static <V,W> Tuple2<Tuple2<Boolean,V>,V> areConnected(WeightedEdge<V,W> edge, Dictionary<V,V> dict){
		V repS = getRep(edge.source(),dict);
		V repD = getRep(edge.destination(),dict);
		return new Tuple2<>(new Tuple2<>(repS.equals(repD),repS), edge.destination());
	}

	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {
		// Creamos el diccionario, la cola de prioridad y nuestro conjunto solución
		Dictionary<V,V> dict = new HashDictionary<>();
		PriorityQueue<WeightedEdge<V,W>> e = new LinkedPriorityQueue<>();
		Set<WeightedEdge<V,W>> sol = new HashSet<>();
		// Ahora inicializo ambos
		// Diccionario
		for(V vertex: g.vertices()){
			dict.insert(vertex,vertex);
		}
		// Cola de prioridad
		for(WeightedEdge<V,W> edge: g.edges()){
			e.enqueue(edge);
		}
		// Tras inicializarlos realizamos el algoritmo
		while(!e.isEmpty()){
			// Sacamos la primera arista
			WeightedEdge<V,W> edge = e.first();
			// Primero miramos que los vértices extraídos de la cola estén conectados
			Tuple2<Tuple2<Boolean,V>,V> t = areConnected(edge,dict);
			if(!t._1()._1()){
				// Actualizo el diccionario
				dict.insert(t._1()._2(),t._2());
				// Añado la arista al bosque de recubrimiento
				sol.insert(edge);
			}
			e.dequeue();
		}
		return sol;
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}

File: ED del Pedro/ED/Haskell/Krustal/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph

fromJust :: Maybe a -> a
fromJust Nothing = error "No tiene valor asociado"
fromJust (Just x) = x

connectedVertex :: (Ord a, Eq a) => WeightedEdge a w -> D.Dictionary a a -> ((Bool, a),a)
connectedVertex (WE v w u) dic = (aux v u dic,u)
    where 
        aux v u dic = (getRep v dic == getRep u dic, getRep v dic)
            where 
                getRep vertex dic
                    | vertex == value = vertex
                    | otherwise = getRep value dic
                        where
                            value = fromJust (D.valueOf vertex dic)

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal g = aux pq dict []
    where 
        pq = foldr (Q.enqueue) (Q.empty) (edges g)
        dict = foldl (\res v -> D.insert v v res) D.empty (vertices g)
        aux pq dict ws
            | Q.isEmpty pq = ws
            | otherwise = if condition then aux pq' dict ws else aux pq' (D.insert rep u dict) ([edge] ++ ws)
                where
                    edge = Q.first pq
                    pq' = Q.dequeue pq
                    ((condition,rep), u) = connectedVertex edge dict
            -- Si están conectados no se hace nada, sino actualizamos dict y lo añadimos ws

-- Solo para los que no tienen evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined
